### DISCUSSION TITLE
Perhaps the EmbeddedId class is not registered

### DISCUSSION THREAD

**Discussion by tatphamit89** [ebean-orm/ebean#3451] [General]
I got this error when I have two primary key in my Entity model.

```
[CreationException: Unable to create injector, see the following errors:

1) [Guice/ErrorInjectingConstructor]: BeanNotRegisteredException: Could not find BeanDescriptor for class models.AdsId. Perhaps the EmbeddedId class is not registered? See https://ebean.io/docs/trouble-shooting#not-registered
  at EbeanDynamicEvolutions.<init>(EbeanDynamicEvolutions.java:42)
  at EbeanDynamicEvolutions.class(EbeanDynamicEvolutions.java:35)
  while locating EbeanDynamicEvolutions
  at EbeanModule.bindings(EbeanModule.java:20):
Binding(class DynamicEvolutions to ConstructionTarget(class EbeanDynamicEvolutions) eagerly)
      \_ installed by: Modules$OverrideModule -> GuiceableModuleConversions$$anon$4
  while locating DynamicEvolutions
  while locating ApplicationEvolutionsProvider
  at EvolutionsModule.<init>(EvolutionsModule.scala:23):
Binding(class ApplicationEvolutions to ProviderConstructionTarget(class ApplicationEvolutionsProvider) eagerly)
      \_ installed by: Modules$OverrideModule -> GuiceableModuleConversions$$anon$4
  while locating ApplicationEvolutions
```

This is my AdsId Class

```
package models;

import jakarta.persistence.Embeddable;

import java.io.Serializable;
import java.util.Objects;

@Embeddable
public final class AdsId implements Serializable {
    public String user_id;
    public Long id;

    // Default constructor
    public AdsId() {}

    // Constructor
    public AdsId(String user_id, Long id) {
        this.user_id = user_id;
        this.id = id;
    }

    // hashCode and equals
    @Override
    public int hashCode() {
        return Objects.hash(user_id, id);
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        AdsId that = (AdsId) obj;
        return Objects.equals(user_id, that.user_id) && Objects.equals(id, that.id);
    }
}

```

This is my main Entity

```
package models;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;
import io.ebean.Finder;
import io.ebean.Model;
import jakarta.persistence.*;
import utils.Utils;

import java.util.Date;
import java.util.List;

@Entity
public class Ads extends Model {

    @EmbeddedId
    public AdsId adsId;
    @JsonProperty("adName")
    public String name;


    @JsonProperty("adAction")
    public String actions;

    @Transient
    public String adStartAt;

    @Transient
    public String adEndAt;

    @JsonIgnore
    public int sentLimit;

    @JsonIgnore
    public int sentCount;

    @JsonIgnore
    public int ocSentLimit;

    @JsonIgnore
    public int ocSentCount;

    @JsonIgnore
    public Date createdAt;

    @JsonIgnore
    @Version
    public Date updatedAt;

    @JsonIgnore
    public Long status;

    @JsonProperty("adTexts")
    public List<AdsText> adsTextList;

    @JsonProperty("adLinks")
    public List<AdsObject> adsObjectList;


    public static Finder<Long, Ads> finder = new Finder<Long, Ads>(Ads.class);


    public String getAdId(){
        if (!Utils.isEmpty(this.adsId.user_id)) {
            return this.adsId.user_id + "_" + this.adsId.id;
        }
        return this.adsId.id.toString();
    }
}


```

Could you help me, Thank you so much, where I am wrong here?

**Comment by rob-bygrave** [comment]
> BeanNotRegisteredException: Could not find BeanDescriptor for class models.AdsId. Perhaps the EmbeddedId class is not registered? See https://ebean.io/docs/trouble-shooting#not-registered

When Ebean starts it needs to find all the classes it is interested in including `@Entity` and `@Embedable`. 
There are 3 ways this is done:

1. Done explicitly by code that we write using:
- [DatabaseBuilder.addClass()](https://ebean.io/apidoc/15/io.ebean.api/io/ebean/DatabaseBuilder.html#addClass(java.lang.Class))
- Older API [DatabaseConfig.addClass()](https://ebean.io/apidoc/15/io.ebean.api/io/ebean/config/DatabaseConfig.html#addClass(java.lang.Class))

2. This is done via Ebean doing classpath scanning.  Ebean will scan the classpath looking for these classes.
3. When using Query Beans, the querybean-generator annotation processor generates source code that will register the entities and embeddable classes


What this error is saying is that `models.AdsId.class` was not registered. One way we could see this error was if we had code to explicitly register all the entity beans and forgot to also register `models.AdsId.class`.

Q: Are you using query beans and hence the io.ebean:querybean-generator annotation processor? If you are we could look in the generated sources code to see if the embedded classes is in there or not etc.  My suspicion is that you are not using query beans.

If you show us the code in `EbeanModule.java` we can probably see if the entity classes are being explicitly registered (and we just forgot the AdsId.class one). I suspect that isn't the approach used here current BUT ... one option is to add code to explicitly register all the entity and embeddable classes (via DatabaseBuilder.addClass() or the older DatabaseConfig.addClass() or the older ServerConfig.addClass()).

If the entity classes are found by classpath scanning, then there seems to be some issue with the classpath. The things to look at are: Is there a test that uses Ebean, how does that initialise Ebean, does that work, is this issue seen only when trying to run the application etc.

**Reply by tatphamit89** [reply]
@rob-bygrave Thank you so much for reply my question.

Maybe I use number 2. classpath scanning default of Ebean. I use Play Framework 3 and use Ebean 15.5.0

I declare in `application.conf` like this:
```

# Ebean Configuration
ebean.default=["models.*"]
ebean.read=["models.read.*"]
```

`AdsId` and `Ads` will be in `models` packages.

What will I need to do more?
```

lazy val root = (project in file("."))
  .enablePlugins(PlayJava, PlayEbean)
//  .enablePlugins(PlayNettyServer).disablePlugins(PlayPekkoHttpServer) // uncomment to use the Netty backend
  .disablePlugins(PlayFilters)
  .settings(
    name := "ssbp-api",
    version := "4.0-SNAPSHOT",
    crossScalaVersions := Seq("2.13.14", "3.3.3"),
    scalaVersion := crossScalaVersions.value.head,
    libraryDependencies ++= Seq(
      guice,
      jdbc,
      "org.apache.commons" % "commons-lang3" % "3.15.0",
      "com.mysql" % "mysql-connector-j" % "9.0.0",
      "software.amazon.awssdk" % "auth" % "2.25.27",
      "software.amazon.awssdk" % "regions" % "2.26.25",
      "software.amazon.awssdk" % "sqs" % "2.26.24",
      "io.ebean" % "ebean" % "15.5.0",
      "org.slf4j" % "slf4j-api" % "2.0.12",
      "com.fasterxml.jackson.core" % "jackson-databind" % "2.17.0", // or a compatible version
      "com.fasterxml.jackson.module" %% "jackson-module-scala" % "2.17.2"

    ),
    (Test / testOptions) := Seq(Tests.Argument(TestFrameworks.JUnit, "-a", "-v")),
    javacOptions ++= Seq("-Xlint:unchecked", "-Xlint:deprecation", "-Werror"), // "-Werror" fix warning latter,
)
```

**Reply by tatphamit89** [reply]
It is working now, Thank you so much, I got mistaken when declared duplicate in my `application.conf`

```
ebean.default=["models.*", "modelsId.*"]
ebean.read=["models.read.*", "modelsId.*"]

```
